File: src/sub/actualstations.c
#include <stdbool.h>
#include <stdio.h>

#define IEQ 658
#define IST 650

extern int nsta, legs, nsp, nstaeff;
extern int knobs[IEQ], istm[IST][IEQ], nactualsta[IST];
extern double sphase[IST][IEQ];
extern char stn[IST][5];
extern bool single_turbo;
extern FILE *fm_ptr;

void actualstations(void) {
    int nobsp[IST] = {0};
    int nobss[IST] = {0};
    int nofreadings = 0;
    int nofreadp = 0;
    int nofreads = 0;

    for (int i = 0; i < nsta; ++i) {
        nactualsta[i] = 0;
        nobsp[i] = 0;
        nobss[i] = 0;
    }

    nstaeff = 0;

    for (int i = 0; i < legs; ++i) {
        for (int k = 0; k < knobs[i]; ++k) {
            int sta = istm[k][i];
            if (sta < 0 || sta >= nsta) {
                continue;
            }
            nactualsta[sta] += 1;
            if (nsp == 2) {
                if (sphase[k][i] == 0.0f) nobsp[sta] += 1;
                if (sphase[k][i] == 1.0f) nobss[sta] += 1;
            }
            nofreadings += 1;
        }
    }

    if (!single_turbo && fm_ptr != NULL) fprintf(fm_ptr, "\n");

    for (int i = 0; i < nsta; ++i) {
        if (nactualsta[i] > 0) {
            if (!single_turbo && fm_ptr != NULL) {
                if (nsp == 2) {
                    fprintf(fm_ptr, " readings for station %.4s : tot=%4d  P:%4d  S:%4d\n",
                            stn[i], nactualsta[i], nobsp[i], nobss[i]);
                } else {
                    fprintf(fm_ptr, " readings for station %.4s :%4d\n", stn[i], nactualsta[i]);
                }
            }
            nstaeff += 1;
        }
        nofreadp += nobsp[i];
        nofreads += nobss[i];
    }

    if (!single_turbo && fm_ptr != NULL) {
        fprintf(fm_ptr, "\n");
        fprintf(fm_ptr, " Total number of stations with readings:%4d\n", nstaeff);
        fprintf(fm_ptr, "\n");
        fprintf(fm_ptr, " Total number of readings: %7d\n", nofreadings);
        fprintf(fm_ptr, " Total number of P readings: %7d\n", nofreadp);
        fprintf(fm_ptr, " Total number of S readings: %7d\n", nofreads);
        fprintf(fm_ptr, "\n");
    }
}

File: src_fortran/sub/actualstations.f
c
c
c
c
c
c
      subroutine ACTUALSTATIONS
c
cek added separate report of P and S phases 20.8.98
c
      implicit none
      include '../inc/vel_com.inc'
      integer i,nofreadings,k,nobsp(ist),nobss(ist),nofreadp,nofreads
c
      do i=1,nsta
         nactualsta(i)=0
         nobsp(i)=0
         nobss(i)=0
      enddo
      nstaeff=0
      nofreadings=0
      do i=1,legs  ! = neqs+nshot
         do k=1,knobs(i)
            nactualsta( istm(k,i) ) = nactualsta( istm(k,i) ) + 1
            if(nsp.eq.2) then
             if(sphase(k,i).eq.0.0) nobsp(istm(k,i))=nobsp(istm(k,i))+1
             if(sphase(k,i).eq.1.0) nobss(istm(k,i))=nobss(istm(k,i))+1
            endif
            nofreadings=nofreadings+1
         enddo
      enddo
      if(.not.single_turbo) write(16,*)
      do i=1,nsta
         if(nactualsta(i).gt.0)then
            if(.not.single_turbo)then
               if(nsp.eq.2) then
               write(16,'('' readings for station '',a4,'' : tot='',
     &                  i4,''  P:'',i4,''  S:'',i4)')
     &                  stn(i),nactualsta(i),nobsp(i),nobss(i)
               else
               write(16,'('' readings for station '',a4,'' :'',i4)')
     &                  stn(i),nactualsta(i)
               endif
            endif
            nstaeff=nstaeff+1
         endif
         nofreadp=nofreadp+nobsp(i)
         nofreads=nofreads+nobss(i)
      enddo
      if(.not.single_turbo)then
         write(16,*)
         write(16,'('' Total number of stations with readings:'',i4)')
     &           nstaeff
         write(16,*)
         write(16,'('' Total number of readings: '',i7)') nofreadings
         write(16,'('' Total number of P readings: '',i7)') nofreadp
         write(16,'('' Total number of S readings: '',i7)') nofreads
         write(16,*)
      endif
c
      RETURN
      end ! of subr. actualstations
